toolchains/symbiflow.py

```diff
@@ -1001,6 +1003,451 @@ def check_env():
+            iob = iob + res['IOB33M_INBUF_EN']
+        if 'IOB33_OUTBUF' in res:
+            iob = iob + res['IOB33_OUTBUF']
+        if 'IOB33_INBUF_EN' in res:
+            iob = iob + res['IOB33_INBUF_EN']
+        if 'RAMB18E1_RAMB18E1' in res:
+            bram = res['RAMB18E1_RAMB18E1']
+        if 'RAMB36E1_RAMB36E1' in res:
+            bram += res['RAMB36E1_RAMB36E1'] * 2
+        if 'PLLE2_ADV_PLLE2_ADV' in res:
+            pll = res['PLLE2_ADV_PLLE2_ADV']
+
+        ret = {
+            "LUT": lut,
+            "DFF": dff,
+            "BRAM": bram,
+            "CARRY": carry,
+            "GLB": None,
+            "PLL": pll,
+            "IOB": iob,
+        }
+        return ret
+
+    def get_yosys_runtimes(self, logfile):
+        runtime_re = 'CPU: user (\d+\.\d+)s system (\d+\.\d+)s'
+        log = dict()
+        with open(logfile, 'r') as fp:
+            for l in fp:
+                m = re.search(runtime_re, l)
+                if not m:
+                    continue
+                time = float(m[1]) + float(m[2])
+                log['synthesis'] = time
+
+                return log
+
+        assert False, "No run time found for yosys."
+
+    def get_nextpnr_runtimes(self, logfile):
+        log = dict()
+
+        placement = 0.0
+        routing = 0.0
+
+        with open(logfile, 'r') as fp:
+            for l in fp:
+                l = l.strip()
+                if len(l) == 0:
+                    continue
+
+                l = l.lstrip("Info: ")
+                heap_placer_string = "HeAP Placer Time: "
+                sa_placer_string = "SA placement time "
+
+                router1_string = "Router1 time "
+                router2_string = "Router2 time "
+
+                if heap_placer_string in l:
+                    time = float(l.lstrip(heap_placer_string).rstrip("s"))
+                    placement += time
+                elif sa_placer_string in l:
+                    time = float(l.lstrip(sa_placer_string).rstrip("s"))
+                    placement += time
+                elif router1_string in l:
+                    time = float(l.lstrip(router1_string).rstrip("s"))
+                    routing += time
+                elif router2_string in l:
+                    time = float(l.lstrip(router2_string).rstrip("s"))
+                    routing += time
+
+        log["place"] = placement
+        log["route"] = routing
+
+        return log
+
+    def add_runtimes(self):
+        """Returns the runtimes of the various steps"""
+
+        yosys_log = os.path.join(self.out_dir, 'yosys.log')
+        nextpnr_log = os.path.join(self.out_dir, 'nextpnr.log')
+
+        synth_times = self.get_yosys_runtimes(yosys_log)
+        impl_times = self.get_nextpnr_runtimes(nextpnr_log)
+
+        for t in synth_times:
+            self.add_runtime(t, synth_times[t])
+        for t in impl_times:
+            self.add_runtime(t, impl_times[t])
```

Comment:
Add these to the parent Nextpnr class as these contain code that may be duplicated across `Nextpnr<Arch>` classes

Reply:
Moved to [NextpnrGeneric](https://github.com/SymbiFlow/fpga-tool-perf/pull/323/files#diff-69ab290f878020ed67b1008338e82df43674e16b1e45733d5f0d8e04b6601141R585-R604)